sdk/rust-examples/logistic-regression/Cargo.toml

```diff
@@ -8,11 +8,10 @@ description = "Logistic regression example, using the Rust rusty-machine ML libr
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-
-pinecone = "*"
-libveracruz = { path = "../../libveracruz" }
+pinecone = "0.2.3"
 rusty-machine = "0.5.4"
 serde = { version = "1.0.3", features = ["derive"] }
+wasi-types = { git = "https://github.com/veracruz-project/wasi-types.git", branch = "veracruz" }
```

Comment:
It seems strange to me that the examples need to include `wasi-types`, this looks like it's only needed for the ErrNo type?

But since we're going through Rust's `fs` library, shouldn't any errors should actually be `io::Error`?

Would it be easier to understand for new users if we just used `u16`s or example-specific Error types throughout the examples?

Reply:
All error from `fs` will automatically convert to ErrNo. Here I only use wasi-type for the error from the pinecone, meaning, any error raised by pinecone will be converted to `ErrNo::proto`. I think a better idea here is to convert any non-fs related error to an error code out of the range of `ErrNo`. I will update this, and remove wasi-type dependency.

Reply:
FWIW, I agree.  I think using `wasi_types` outside of any code that interacts *directly* with the ABI is strange.

Reply:
I do not think it is minor first of all :).

we are effectively doing `int main()` but the `int` is in the range of `ErrNo`, where each number have a particular meaning defined in the `ErrNo`. If it makes sense?

Reply:
It is something we can change in just the examples later if we want to go ahead and get the PR in.

Returning `ErrNo` from `int main()` works, but is that what we want users to do? Using it in the examples implies Veracruz programs must return `ErrNo`, and most users probably won't think twice about it.

But using `ErrNo` for applications is a bit limiting, ties down the changes we can make to wasi-types, and adds extra noise to the examples.

Reply:
It will also require some change on the engine. As engine expected a `int` that can be pared to ErrNo. I think we can weaken it by allow any `int` but still reserved the lower range for `ErrNo`.

Reply:
Oh does the engine change behavior based on the main/exit return code? I assumed it was just passed to the user.

Reply:
From just a brief scan, it looks like the only behavior change is on `ErrNo::Success` (`0`), which could be checked without assuming the type is an `ErrNo` (and would match traditional `int main()` behavior)

Reply:
The engine just passes the value back to runtime and then the clients. My thought process here is to always pass strongly-typed value but limit the use of, e.g. `u32`. I personally feel the latter is prone to error.

Reply:
It could return a `Result<(), NonZeroU32>`. Is it possible for `ErrNo` to get returned from `invoke_entry_point` another way? There is also the `FatalEngineError` which looks like it covers the engine itself failing.

Which I guess would make the full type `Result<Result<(), NonZeroU32>, FatalEngineError>`.

Reply:
It is possible to return anything in `invoke_entry_point` since we control it. Yet it makes sense for wasm program to return`ErrNo` back to the execution engine interface, namely `invoke_entry_point`, and the execution engine should not treat `ErrNo` as a fatal error (Subject to debate). 

`Result<(),NonZeroU32>` in the invoke_entry_point has no problem at all. Yet I tried `Result<(),NonZeroU32>` in the `main` in examples. I believe the wasm program complied from rust will try to write the value of `NonZeroU32` to the stderr rather than return an `int` to the callee, it makes sense to me as WASM is value-based, and you cannot guarantee what is in the Error side of the result type, so rust uniformly compiles error to write to stderr. 

I think the census here is to allow returning arbitrary int, and the execution engine does not make any assumption on it as it is return code. I will update it.

Reply:
Ah yeah, I suspect trying to change the return type of `main` will unfortunately make the stdlibs quite unhappy

Reply:
Just updated. Remove the ErrNo dependency in the example and downgrade the return code as an `i32` hence the return code for invoke_entry_point to `i32`.